### PR TITLE
8339242: Fix overflow issues in AdlArena

### PR DESCRIPTION
**Pull request by caspernorrbin**
Hi everyone,

This PR addresses an issue in `adlArena` where some allocations lack checks for overflow. This could potentially result in successful allocations when called with unrealistic values.

The fix includes:

- Adding assertions to check for potential overflow.
- Reordering some operations to guard against overflow.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339242](https://bugs.openjdk.org/browse/JDK-8339242): Fix overflow issues in AdlArena (**Bug** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20774/head:pull/20774` \
`$ git checkout pull/20774`

Update a local copy of the PR: \
`$ git checkout pull/20774` \
`$ git pull https://git.openjdk.org/jdk.git pull/20774/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20774`

View PR using the GUI difftool: \
`$ git pr show -t 20774`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20774.diff">https://git.openjdk.org/jdk/pull/20774.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20774#issuecomment-2320544733)

### CODE REVIEW

File: src/hotspot/share/adlc/adlArena.cpp [hunk 1]
```diff
@@ -147,8 +149,9 @@ void *AdlArena::Arealloc( void *old_ptr, size_t old_size, size_t new_size ) {
   }
 
   // See if we can resize in-place
-  if( (c_old+old_size == _hwm) &&       // Adjusting recent thing
-      (c_old+new_size <= _max) ) {      // Still fits where it sits
+  if( (c_old+old_size == _hwm) &&            // Adjusting recent thing
+      ((size_t)(_max-c_old) >= new_size) ) { // Still fits where it sits, safe from overflow
+
```
Comment (dean-long) [inline src/hotspot/share/adlc/adlArena.cpp:154]: This code appears to be a copy of Arena::Arealloc, so we should probably fix both at the same time.

**Comment by tstuefe** [comment]
If the aim is to increase security, would it not make more sense to test against hardcoded "reasonable max" values? Anything larger than a few MB is likely to be an error anyway, or?

**Comment by jdksjolen** [comment]
> If the aim is to increase security, would it not make more sense to test against hardcoded "reasonable max" values? Anything larger than a few MB is likely to be an error anyway, or?

This is a build-time only feature, as it's run on `*.ad` files, so I think this should count as trusted input. The assertions in the PR just let's us bail early rather than discovering the overflow later on through a crash (or not at all).

**Review by jdksjolen — APPROVED** [review]
LGTM

File: src/hotspot/share/adlc/adlArena.cpp (hunk 1 above)
Comment (caspernorrbin) [reply src/hotspot/share/adlc/adlArena.cpp:154]: Good catch! Fixed in cf0b434.

File: src/hotspot/share/memory/arena.cpp [hunk 2]
```diff
@@ -335,8 +335,8 @@ void *Arena::Arealloc(void* old_ptr, size_t old_size, size_t new_size, AllocFail
   size_t corrected_new_size = ARENA_ALIGN(new_size);
 
   // See if we can resize in-place
-  if( (c_old+old_size == _hwm) &&       // Adjusting recent thing
-      (c_old+corrected_new_size <= _max) ) {      // Still fits where it sits
+  if( (c_old+old_size == _hwm) &&                      // Adjusting recent thing
+      ((size_t)(_max-c_old) >= corrected_new_size) ) { // Still fits where it sits, safe from overflow
```
Comment (tstuefe) [inline src/hotspot/share/memory/arena.cpp:339]: This change is correct, but it hides an important finding behind a reshuffling of parameters that someone else may innocently reshape later. And in the process, makes the code less readable. Can we use something like saturated_add()?

I would also add an explicit assert for a reasonable max size. Arena allocations should be small. Nobody should hand in sizes larger than a few MB, so asserting for size >= 2^31 (2g) would make sense. Anything as large as that is almost certainly an error we should trap on.

Comment (caspernorrbin) [reply src/hotspot/share/memory/arena.cpp:339]: I would support this.

May be possible to just use `saturated_add()`. Would be limited to `max_jint`, but with an assert for sizes above 2^31, that shouldn't be an issue. Will look into it.

Comment (kimbarrett) [reply src/hotspot/share/memory/arena.cpp:339]: Please, not `max_jint`.  This limit isn't tied to Java integer semantics.  `MAX_INT` might be appropriate.
(Yes, I know they likely have the same value.)

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/memory/arena.cpp (hunk 2 above)
Comment (caspernorrbin) [reply src/hotspot/share/memory/arena.cpp]: @tstuefe I've now added asserts and changed this check to use saturated adds.

The already existing `saturated_add()` only works on `int`s and `uint`s, so I made a version working on offsetting pointers. We still have to check for overflow (if the pointer addition was saturated), but the code becomes more readable.

File: src/hotspot/share/adlc/adlArena.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (caspernorrbin) [reply src/hotspot/share/adlc/adlArena.cpp:165]: I believe this would miss the case where we shrink an allocation in place and we are not at the high water mark, where `new_size <= old_size`, but where `c_old + old_size) == _hwm` does not hold.

[the remark this answers is not in the document]
Comment (kimbarrett) [reply src/hotspot/share/adlc/adlArena.cpp:165]: Oops, I managed to drop some of the code when pasting it into the github
comment. There should be an `else if` clause:
```
if ((c_old + old_size) == _hwm) {
  ...
} else if (new_size <= old_size) {
  return c_old;
}
```
I like this version better than the earlier version because it has only one
copy of the _hwm handling. It trades an additional pointer-delta in the
shrink-last-allocation case for avoiding a compare-and-branch in the
grow-last-allocation case, which I think is a good tradeoff, though likely
pretty minor.

I do not like the new saturated_add version.  I think that just makes things
(slightly) slower vs the pointer_delta compare, with no other benefit.

File: src/hotspot/share/adlc/adlArena.hpp [hunk 3]
```diff
@@ -104,9 +104,12 @@ class AdlArena: public AdlCHeapObj {
 
   // Fast allocate in the arena.  Common case is: pointer test + increment.
   void* Amalloc(size_t x) {
+  assert(x <= (size_t)1 << 31, "unreasonable arena allocation size");
```
Comment (kimbarrett) [inline src/hotspot/share/adlc/adlArena.hpp:107]: This seems mostly unrelated to the other overflow avoidance checking, and I think doesn't belong in
this PR.  For one thing, this might not be the right mechanism.  Perhaps an Arena should have a
configurable max allocation size?  And shouldn't the max allocation be applied to Arealloc too?

File: src/hotspot/share/adlc/adlArena.cpp [hunk 4]
```diff
@@ -129,28 +130,41 @@ void* AdlArena::grow( size_t x ) {
...
 // Reallocate storage in AdlArena.
 void *AdlArena::Arealloc( void *old_ptr, size_t old_size, size_t new_size ) {
   char *c_old = (char*)old_ptr; // Handy name
   // Stupid fast special case
-  if( new_size <= old_size ) {  // Shrink in-place
-    if( c_old+old_size == _hwm) // Attempt to free the excess bytes
-      _hwm = c_old+new_size;    // Adjust hwm
+  if (new_size <= old_size) {     // Shrink in-place
+    if (c_old + old_size == _hwm) // Attempt to free the excess bytes
+      _hwm = c_old + new_size;    // Adjust hwm
```
Comment (kimbarrett) [inline src/hotspot/share/adlc/adlArena.cpp:156]: This `if` is missing braces around the consequent.  If fixing the whitespace, the missing braces should
also be added.

**Review by kimbarrett — CHANGES_REQUESTED** [review]
I've only commented on the adlc changes, but I think all of those comments also
apply to Arena.

**Comment by caspernorrbin** [comment]
Here are my thoughts after looking through this again and reading your comments.

I agree that blindly reshuffling parameters to prevent overflow may not be the best solution. I think using something like `pointer_delta` strikes a good balance.  Using a saturated add feels to me like checking for overflow, just with extra steps.

I also believe that asserting for ”reasonable” arena allocation sizes may be out of scope of this PR. The initial purpose was to fix potential overflow issues in the adlc arena, and has already expanded into the regular arena. These asserts also causes test failures, so even more changes would be required if added. It may be better to create a separate issue for generally improving arena allocation safety.

With the overflow checks in place, Kim’s refactor is a bit cleaner and more readable than the modified original, so I’ve opted to implement that.


Please let me know any further comments or suggestions

File: src/hotspot/share/adlc/adlArena.cpp [hunk 5]
```diff
@@ -129,28 +130,33 @@ void* AdlArena::grow( size_t x ) {
...
 void *AdlArena::Acalloc( size_t items, size_t x ) {
+  assert(items <= SIZE_MAX / x, "overflow");
   size_t z = items*x;   // Total size needed
   void *ptr = Amalloc(z);       // Get space
   memset( ptr, 0, z );          // Zap space
   return ptr;                   // Return space
 }
 
 //------------------------------realloc----------------------------------------
+size_t pointer_delta(const void *left, const void *right) {
```
Comment (dean-long) [inline src/hotspot/share/adlc/adlArena.cpp:141]: Do we want to assert left >= right here?

Comment (kimbarrett) [inline src/hotspot/share/adlc/adlArena.cpp:141]: Function ought to be static.

Comment (kimbarrett) [reply src/hotspot/share/adlc/adlArena.cpp:141]: It's currently only used in one place, where we know that's true.  OTOH, an assert doesn't hurt.
Instead of this helper function, we could just use`(size_t)(_max - c_old)` inline (we can be
confident the difference won't exceed the `ptrdiff_t` range here), reducing the lines of code by
a little bit.  The benefit from the helper is having the Arealloc code (nearly? completely?) identical
here and in memory/arena.cpp.  I don't have a strong opinion either way.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/adlc/adlArena.cpp (hunk 5 above)
Comment (caspernorrbin) [reply src/hotspot/share/adlc/adlArena.cpp]: The motivation was to keep the two arenas consistent. To that end, I've now added an assert and made the function static.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by jdksjolen — APPROVED** [review]
Still looks good to me, ship it!

**Comment by caspernorrbin** [comment]
Thank you everyone for the discussion and the reviews.

/integrate
